### PR TITLE
PP-10835: Remove webhook messages after a configurable number of days

### PR DESCRIPTION
**Pull request by oswaldquek**
As we want to remove webhooks messages based on a maximum age of message,
introduce methods to retrieve the WebhookMessageEntities and
WebhookDeliveryQueueEntities based on date, and methods to delete them.

### CODE REVIEW

File: src/main/java/uk/gov/pay/webhooks/app/WebhookMessageDeletionConfig.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package uk.gov.pay.webhooks.app;
+
+import javax.validation.Valid;
+import javax.validation.constraints.NotNull;
+
+public class WebhookMessageDeletionConfig {
+    
+    @Valid
+    @NotNull
+    private int maxAgeOfMessages;
```
Comment (nlsteers) [inline src/main/java/uk/gov/pay/webhooks/app/WebhookMessageDeletionConfig.java:10]: looks like the `config.yaml` entry for these vars is missing

File: src/test/java/uk/gov/pay/webhooks/webhook/WebhookServiceTest.java [hunk 2]
```diff
@@ -43,7 +51,49 @@ class WebhookServiceTest {
...
+        when(webhookMessageDeletionConfig.getMaxAgeOfMessages()).thenReturn(7);
+        when(webhookMessageDeletionConfig.getMaxNumOfMessagesToExpire()).thenReturn(2);
+        
+        var webhook = new WebhookEntity();
+        var webhookMessage1 = createWebhookMessageEntity(webhook);
+        var webhookMessage2 = createWebhookMessageEntity(webhook);
+        var webhookMessage3 = createWebhookMessageEntity(webhook);
+        var webhookMessage4 = createWebhookMessageEntity(webhook);
+        when(webhookMessageDao.getWebhookMessagesOlderThan(7)).thenReturn(
+                List.of(webhookMessage1, webhookMessage2, webhookMessage3, webhookMessage4));
```
Comment (nlsteers) [inline src/test/java/uk/gov/pay/webhooks/webhook/WebhookServiceTest.java:69]:
```suggestion
        when(webhookMessageDao.getWebhookMessagesOlderThan(7)).thenReturn(
                List<WebhookMessageEntity> webhookMessages = Collections.nCopies(4, createWebhookMessageEntity(webhook)));
```
could DRY this test up a bit?

File: src/test/java/uk/gov/pay/webhooks/webhook/WebhookServiceTest.java [hunk 3]
```diff
@@ -43,7 +51,49 @@ class WebhookServiceTest {
...
+
+        var webhookDeliveryQueueEntity1 = new WebhookDeliveryQueueEntity();
+        var webhookDeliveryQueueEntity2 = new WebhookDeliveryQueueEntity();
+        var webhookDeliveryQueueEntity3 = new WebhookDeliveryQueueEntity();
+        var webhookDeliveryQueueEntity4 = new WebhookDeliveryQueueEntity();
+        when(webhookDeliveryQueueDao.getWebhookDeliveryQueueEntitiesOlderThan(7)).thenReturn(
```
Comment (nlsteers) [inline src/test/java/uk/gov/pay/webhooks/webhook/WebhookServiceTest.java:75]: same here

File: src/test/java/uk/gov/pay/webhooks/message/dao/WebhookMessageDaoTest.java [hunk 4]
```diff
@@ -40,6 +41,32 @@ public void setUp() {
         webhookDeliveryQueueDao = new WebhookDeliveryQueueDao(database.getSessionFactory(), InstantSource.fixed(Instant.now()));
     }
 
+    @Test
+    public void shouldDeleteWebhookMessages() {
```
Comment (nlsteers) [inline src/test/java/uk/gov/pay/webhooks/message/dao/WebhookMessageDaoTest.java:45]: this test has no assertions

File: src/test/java/uk/gov/pay/webhooks/message/dao/WebhookMessageDaoTest.java [hunk 5]
```diff
@@ -40,6 +41,32 @@ public void setUp() {
...
+        database.inTransaction(() -> {
+            var webhook = new WebhookEntity();
+            var webhookMessage1 = createWebhookMessageEntity(webhook);
+            var webhookMessage2 = createWebhookMessageEntity(webhook);
+            var webhookMessage3 = createWebhookMessageEntity(webhook);
+            webhookMessageDao.deleteMessages(Stream.of(webhookMessage1, webhookMessage2, webhookMessage3));
+        });
+    }
+
+    private WebhookMessageEntity createWebhookMessageEntity(WebhookEntity webhook) {
```
Comment (nlsteers) [inline src/test/java/uk/gov/pay/webhooks/message/dao/WebhookMessageDaoTest.java:55]: maybe move this down to the other private methods

File: src/main/java/uk/gov/pay/webhooks/app/WebhookMessageDeletionConfig.java (hunk 1 above)
Comment (oswaldquek) [reply src/main/java/uk/gov/pay/webhooks/app/WebhookMessageDeletionConfig.java:10]: It's not really needed just yet as the default values are 0 and there's no endpoint to trigger the `WebhookService.deleteWebhookMessagesOlderThan` just yet. The important thing is that the integration tests pass.

File: src/test/java/uk/gov/pay/webhooks/message/dao/WebhookMessageDaoTest.java (hunk 5 above)
Comment (oswaldquek) [reply src/test/java/uk/gov/pay/webhooks/message/dao/WebhookMessageDaoTest.java]: fixed

File: src/main/java/uk/gov/pay/webhooks/app/WebhookMessageDeletionConfig.java (hunk 1 above)
Comment (nlsteers) [reply src/main/java/uk/gov/pay/webhooks/app/WebhookMessageDeletionConfig.java:10]: no worries, just weren't sure if you were implementing the second part of this story!

**Pull request merged**